### PR TITLE
Add jenkins workload from statefulSet 

### LINKED ISSUE #45: Turn Jenkins workload from deployment to statefulset
The [official Jenkins helm chart](https://github.com/jenkinsci/helm-charts/blob/main/charts/jenkins/templates/jenkins-controller-statefulset.yaml) uses statefulset instead of deployment. I believe it's reasonable.

* Jenkins does not allow two instances to use the same storage. So the replicas of Jenkins deploy cannot be bigger than 1.

## Compitable

Please don't forget the upgrade process. I mean that please make sure users can upgrade from the deployment style to statefulset way successfully.

## References
* https://helm.sh/
* [Kubernetes Deployments vs StatefulSets](https://stackoverflow.com/questions/41583672/kubernetes-deployments-vs-statefulsets)

### PR DESCRIPTION
**Pull request by shihaoH**
Fix #45 
description : 
The StatefulSet workload can be enabled by configuring `jenkins.Master.StatefulSet.Enabled`. 
And statefulSet volumeClaimTemplates can be enabled by turning off `jenkins.Persistence.Enable` and configuring `jenkins.Persistence.ExistingClaim`.

### CODE REVIEW

File: charts/ks-devops/charts/jenkins/values.yaml [hunk 1]
```diff
@@ -1,4 +1,11 @@
 Master:
+  StatefulSet:
+    Enable: true
+    Replicas: 1
+    # StatefulSetUpdateStrategy must be 'RollingUpdate' or 'OnDelete'
+    UpdateStrategy: RollingUpdate
```
Comment (JohnNiang) [inline charts/ks-devops/charts/jenkins/values.yaml:6]: `OnDelete` may be reasonable, because there is only one Jenkins instance forever.

**Comment by JohnNiang** [comment]
Thanks for your time.

**Comment by shihaoH** [comment]
> Thanks for your time. BTW, `issue: #45` in PR description should be `Fix #45`. See also [here](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

Is it right ?

**Comment by LinuxSuRen** [comment]
> > Thanks for your time. BTW, `issue: #45` in PR description should be `Fix #45`. See also [here](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
> 
> Is it right ?

Yes. See 
image

**Review by LinuxSuRen — CHANGES_REQUESTED** [review]

**Comment by LinuxSuRen** [comment]
The pipeline cannot run after I installed the Jenkins chart with your PR.

My sample Jenkinsfile:
```
pipeline {
    agent {
        label 'go'
    }

    stages {
        stage('Hello') {
            steps {
                echo 'Hello World'
                sh '''
                go version
                '''
            }
        }
    }
}
```

I can find the following error output from Jenkins:
```
Error in provisioning; agent=KubernetesSlave name: go-k38pv, template=PodTemplate{id='1b946001-072a-4d82-9bbf-92213c3a89da', name='go', namespace='kubesphere-devops-worker', label='go', nodeUsageMode=EXCLUSIVE, workspaceVolume=EmptyDirWorkspaceVolume [memory=false], volumes=[HostPathVolume [mountPath=/var/run/docker.sock, hostPath=/var/run/docker.sock], HostPathVolume [mountPath=/home/jenkins/go/pkg, hostPath=/var/data/jenkins_go_cache], HostPathVolume [mountPath=/root/.sonar/cache, hostPath=/var/data/jenkins_sonar_cache]], containers=[ContainerTemplate{name='go', image='kubespheredev/builder-go:v3.1.0', command='cat', args='', ttyEnabled=true, resourceRequestCpu='100m', resourceRequestMemory='100Mi', resourceLimitCpu='4000m', resourceLimitMemory='8192Mi'}, ContainerTemplate{name='jnlp', image='jenkins/jnlp-slave:3.27-1', command='jenkins-slave', args='${computer.jnlpmac} ${computer.name}', resourceRequestCpu='50m', resourceRequestMemory='400Mi', resourceLimitCpu='500m', resourceLimitMemory='1536Mi'}]}
Also:   java.lang.Throwable: launched here
	at hudson.slaves.SlaveComputer._connect(SlaveComputer.java:283)
	at hudson.model.Computer.connect(Computer.java:435)
	at hudson.slaves.CloudRetentionStrategy.start(CloudRetentionStrategy.java:73)
	at org.jenkinsci.plugins.durabletask.executors.OnceRetentionStrategy.start(OnceRetentionStrategy.java:83)
	at org.jenkinsci.plugins.durabletask.executors.OnceRetentionStrategy.start(OnceRetentionStrategy.java:46)
	at hudson.model.AbstractCIBase.updateComputer(AbstractCIBase.java:162)
	at hudson.model.AbstractCIBase.access$000(AbstractCIBase.java:44)
	at hudson.model.AbstractCIBase$2.run(AbstractCIBase.java:224)
	at hudson.model.Queue._withLock(Queue.java:1398)
	at hudson.model.Queue.withLock(Queue.java:1275)
	at hudson.model.AbstractCIBase.updateComputerList(AbstractCIBase.java:207)
	at jenkins.model.Jenkins.updateComputerList(Jenkins.java:1635)
	at jenkins.model.Nodes$2.run(Nodes.java:139)
	at hudson.model.Queue._withLock(Queue.java:1398)
	at hudson.model.Queue.withLock(Queue.java:1275)
	at jenkins.model.Nodes.addNode(Nodes.java:135)
	at jenkins.model.Jenkins.addNode(Jenkins.java:2158)
	at hudson.slaves.NodeProvisioner.lambda$update$6(NodeProvisioner.java:256)
	at hudson.model.Queue._withLock(Queue.java:1398)
	at hudson.model.Queue.withLock(Queue.java:1275)
	at hudson.slaves.NodeProvisioner.update(NodeProvisioner.java:225)
	at hudson.slaves.NodeProvisioner.lambda$suggestReviewNow$1(NodeProvisioner.java:183)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://kubernetes.default/api/v1/namespaces/kubesphere-devops-worker/pods. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. admission webhook "resourcesquotas.quota.kubesphere.io" denied the request: pods "go-k38pv" is forbidden: exceeded quota: worker-resource-quota, requested: limits.cpu=4500m,limits.memory=9728Mi, used: limits.cpu=0,limits.memory=0, limited: limits.cpu=3,limits.memory=3Gi.
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:589)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:526)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:492)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:451)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:252)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:841)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:332)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:82)
	at org.csanchez.jenkins.plugins.kubernetes.KubernetesLauncher.launch(KubernetesLauncher.java:133)
	at hudson.slaves.SlaveComputer.lambda$_connect$0(SlaveComputer.java:294)
	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
	at jenkins.security.ImpersonatingExecutorService$2.call(ImpersonatingExecutorService.java:71)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by JohnNiang** [comment]
hi @LinuxSuRen , it's my fault, not his problem. please see also: <https://github.com/kubesphere-sigs/ks-devops/issues/73>.

**Review by LinuxSuRen — COMMENTED** [review]
/lgtm

**Review by LinuxSuRen — COMMENTED** [review]
Thanks for your contribution.

File: charts/ks-devops/charts/jenkins/values.yaml (hunk 1 above)
Comment (JohnNiang) [reply charts/ks-devops/charts/jenkins/values.yaml:6]: @LinuxSuRen what do you reckon to this strategy.

Comment (shihaoH) [reply charts/ks-devops/charts/jenkins/values.yaml:6]: If this stategy is set to "OnDelete", an error will occur when executing "helm upgrade"

Comment (JohnNiang) [reply charts/ks-devops/charts/jenkins/values.yaml:6]: Okay, I will test against it.

**Comment by JohnNiang** [comment]
hi @shihaoH , please update your PR description: `Enable` to `Enabled`.

/lgtm

**Pull request merged**